azure-kusto-data/azure/kusto/data/client.py

```diff
@@ -584,15 +599,56 @@ def __init__(self, kcsb: Union[KustoConnectionStringBuilder, str]):
 
         # Create a session object for connection pooling
         self._session = requests.Session()
-        self._session.mount("http://", HTTPAdapter(pool_maxsize=self._max_pool_size))
-        self._session.mount("https://", HTTPAdapter(pool_maxsize=self._max_pool_size))
+        adapter = HTTPAdapterWithSocketOptions(socket_options=self.compose_socket_options(), pool_maxsize=self._max_pool_size, max_retries=retries)
+        self._session.mount("http://", adapter)
+        self._session.mount("https://", adapter)
+
         self._mgmt_endpoint = "{0}/v1/rest/mgmt".format(kusto_cluster)
         self._query_endpoint = "{0}/v2/rest/query".format(kusto_cluster)
         self._streaming_ingest_endpoint = "{0}/v1/rest/ingest/".format(kusto_cluster)
         # notice that in this context, federated actually just stands for add auth, not aad federated auth (legacy code)
         self._auth_provider = _AadHelper(kcsb) if kcsb.aad_federated_security else None
         self._request_headers = {"Accept": "application/json", "Accept-Encoding": "gzip,deflate", "x-ms-client-version": "Kusto.Python.Client:" + VERSION}
 
+    @staticmethod
+    def compose_socket_options():
+        # Sends TCP Keep-Alive after MAX_IDLE_SECONDS seconds of idleness, once every INTERVAL_SECONDS seconds, and closes the connection after MAX_FAILED_KEEPALIVES failed pings (e.g. 20 => 1:00:30)
+        MAX_IDLE_SECONDS = 30

```

Comment:
What are the default values and why do we need to change them? If we can let's keep the default values and save the different OS's settings.

Reply:
I'll research this further, but the problem is that it's not really documented. As far as I understand, we need to set all values, including the timeouts, for it to work (no default).

But even if there were defaults, I'd think it'd be best to set consistent Keep-Alive timings across all OSes, and deliberately for what we want. Why use an OS default?

Reply:
I researched and tested, and indeed there are no defaults for many of these. For example, on Windows if you just set:
`(socket.SOL_SOCKET, socket.SO_KEEPALIVE, 1)`
But not:
`(socket.IPPROTO_TCP, socket.TCP_KEEPIDLE, [number of seconds])`
Then no TCP Keep-Alives will be sent.